File: lib/java/opentoken/src/main/java/com/truveta/opentoken/tokentransformer/JweMatchTokenFormatter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.tokentransformer;

import com.nimbusds.jose.EncryptionMethod;
import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWEAlgorithm;
import com.nimbusds.jose.JWEHeader;
import com.nimbusds.jose.JOSEObjectType;
import com.nimbusds.jose.JWEObject;
import com.nimbusds.jose.Payload;
import com.nimbusds.jose.crypto.DirectEncrypter;
import com.nimbusds.jose.jwk.OctetSequenceKey;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Formats tokens in the JWE-based match token format (ot.V1.&lt;JWE&gt;).
 * <p>
 * This transformer wraps the privacy-protected identifier (PPID) in a 
 * self-contained JWE structure with all necessary metadata for versioning
 * and cryptographic agility.
 * 
 * @see <a href="https://datatracker.ietf.org/doc/html/rfc7516">RFC 7516 - JSON Web Encryption (JWE)</a>
 */
public class JweMatchTokenFormatter implements TokenTransformer {
    private static final long serialVersionUID = 1L;
    private static final Logger logger = LoggerFactory.getLogger(JweMatchTokenFormatter.class);

    private static final String TOKEN_PREFIX = "ot.V1.";
    private static final String TOKEN_TYPE = "match-token";

    private final String ringId;
    private final String ruleId;
    private final String issuer;
    private final DirectEncrypter encrypter;

    /**
     * Initializes the JWE match token formatter.
     * 
     * @param encryptionKey the encryption key (must be 32 bytes for AES-256)
     * @param ringId the ring identifier for key management
     * @param ruleId the token rule identifier (e.g., "T1", "T2", etc.)
     * @param issuer the issuer identifier (optional, defaults to "truveta.opentoken")
     * @throws JOSEException if the encrypter cannot be initialized
     */
    public JweMatchTokenFormatter(String encryptionKey, String ringId, String ruleId, String issuer)
            throws JOSEException {
        if (encryptionKey == null || encryptionKey.length() != 32) {
            throw new IllegalArgumentException("Encryption key must be exactly 32 characters (256 bits)");
        }
        if (ringId == null || ringId.isEmpty()) {
            throw new IllegalArgumentException("Ring ID must not be null or empty");
        }
        if (ruleId == null || ruleId.isEmpty()) {
            throw new IllegalArgumentException("Rule ID must not be null or empty");
        }

        this.ringId = ringId;
        this.ruleId = ruleId;
        this.issuer = (issuer != null && !issuer.isEmpty()) ? issuer : "truveta.opentoken";

        // Create a 256-bit key from the encryption key
        byte[] keyBytes = encryptionKey.getBytes(StandardCharsets.UTF_8);
        OctetSequenceKey jwk = new OctetSequenceKey.Builder(keyBytes).build();

        // Initialize direct encrypter for AES-256-GCM
        this.encrypter = new DirectEncrypter(jwk);
    }

    /**
     * Transforms a token (PPID) into the JWE match token format.
     * <p>
     * The input token should be the base64-encoded HMAC output from previous transformers.
     * This method wraps it in a JWE structure with metadata and prepends the "ot.V1." prefix.
     * 
     * @param token the privacy-protected identifier (PPID) to wrap in JWE format
     * @return the formatted match token: ot.V1.&lt;JWE compact serialization&gt;
     * @throws Exception if JWE encryption or serialization fails
     */
    @Override
    public String transform(String token) throws Exception {
        if (token == null || token.isEmpty()) {
            logger.warn("Received null or empty token for rule {}", ruleId);
            return token; // Return as-is for blank tokens
        }

        try {
            // Build the JWE payload with metadata using a Map
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("rlid", ruleId);
            payload.put("hash_alg", "SHA-256");
            payload.put("mac_alg", "HS256");
            payload.put("ppid", Collections.singletonList(token));
            payload.put("rid", ringId);
            payload.put("iss", issuer);
            payload.put("iat", Instant.now().getEpochSecond());

            // Create JWE header with algorithm and encryption method
            JWEHeader header = new JWEHeader.Builder(JWEAlgorithm.DIR, EncryptionMethod.A256GCM)
                    .type(new JOSEObjectType(TOKEN_TYPE))
                    .keyID(ringId)
                    .build();

            // Create JWE object with the payload (Payload accepts Map and converts to JSON)
            JWEObject jweObject = new JWEObject(header, new Payload(payload));

            // Encrypt the JWE object
            jweObject.encrypt(encrypter);

            // Serialize to compact form and prepend the ot.V1. prefix
            String jweCompact = jweObject.serialize();
            return TOKEN_PREFIX + jweCompact;

        } catch (JOSEException e) {
            logger.error("Failed to create JWE token for rule {}: {}", ruleId, e.getMessage());
            throw new Exception("JWE token generation failed", e);
        }
    }
}

File: lib/python/opentoken/src/main/opentoken/tokentransformer/jwe_match_token_formatter.py
"""
Copyright (c) Truveta. All rights reserved.

JWE Match Token Formatter for OpenToken V1 format.
"""

import base64
import json
import time
from typing import Optional
from jwcrypto import jwe, jwk
from opentoken.tokentransformer.match_token_constants import V1_TOKEN_PREFIX
from opentoken.tokentransformer.token_transformer import TokenTransformer


class JweMatchTokenFormatter(TokenTransformer):
    """
    Formats tokens in the JWE-based match token format (ot.V1.<JWE>).
    
    This formatter wraps the privacy-protected identifier (PPID) in a 
    self-contained JWE structure with all necessary metadata for versioning
    and cryptographic agility.
    
    See RFC 7516 - JSON Web Encryption (JWE)
    """
    
    TOKEN_TYPE = "match-token"
    
    def __init__(self, encryption_key: str, ring_id: str, rule_id: str, issuer: Optional[str] = None):
        """
        Initialize the JWE match token formatter.
        
        Args:
            encryption_key: The encryption key (must be 32 bytes for AES-256)
            ring_id: The ring identifier for key management
            rule_id: The token rule identifier (e.g., "T1", "T2", etc.)
            issuer: The issuer identifier (optional, defaults to "truveta.opentoken")
            
        Raises:
            ValueError: If encryption_key, ring_id, or rule_id are invalid
        """
        if not encryption_key or len(encryption_key) != 32:
            raise ValueError("Encryption key must be exactly 32 characters (256 bits)")
        if not ring_id:
            raise ValueError("Ring ID must not be None or empty")
        if not rule_id:
            raise ValueError("Rule ID must not be None or empty")
        
        self.ring_id = ring_id
        self.rule_id = rule_id
        self.issuer = issuer if issuer else "truveta.opentoken"
        
        # Create JWK from the encryption key - needs to be base64url-encoded
        key_bytes = encryption_key.encode('utf-8')
        key_b64 = base64.urlsafe_b64encode(key_bytes).decode('utf-8').rstrip('=')
        self.jwk_key = jwk.JWK(kty="oct", k=key_b64)
    
    def transform(self, token: str) -> str:
        """
        Transform a token (PPID) into the JWE match token format.
        
        The input token should be the base64-encoded HMAC output from previous transformers.
        This method wraps it in a JWE structure with metadata and prepends the "ot.V1." prefix.
        
        Args:
            token: The privacy-protected identifier (PPID) to wrap in JWE format
            
        Returns:
            The formatted match token: ot.V1.<JWE compact serialization>
            
        Raises:
            Exception: If JWE encryption or serialization fails
        """
        if not token:
            # Return as-is for blank tokens
            return token
        
        try:
            # Build the JWE payload with metadata
            payload = {
                "rlid": self.rule_id,
                "hash_alg": "SHA-256",
                "mac_alg": "HS256",
                "ppid": [token],  # PPID as an array (single element for hash-based tokens)
                "rid": self.ring_id,
                "iss": self.issuer,
                "iat": int(time.time())
            }
            
            # Create JWE header with algorithm and encryption method
            protected_header = {
                "alg": "dir",  # Direct encryption (key used directly)
                "enc": "A256GCM",  # AES-256-GCM encryption
                "typ": self.TOKEN_TYPE,
                "kid": self.ring_id
            }
            
            # Create JWE object
            jwe_token = jwe.JWE(
                plaintext=json.dumps(payload).encode('utf-8'),
                recipient=self.jwk_key,
                protected=protected_header
            )
            
            # Serialize to compact form and prepend the ot.V1. prefix
            jwe_compact = jwe_token.serialize(compact=True)
            return V1_TOKEN_PREFIX + jwe_compact
            
        except Exception as e:
            raise Exception(f"JWE token generation failed for rule {self.rule_id}: {str(e)}")
